File: samples/messy.py
# A deliberately awful Python file, to prove Chewgy is language-agnostic.

import os
import json
import subprocess

API_KEY = "sk-live-please-do-not-commit-this-0987654321"


# Over-engineering: an abstract factory for two constants.
class AbstractGreeterFactoryBase:
    def create(self, lang):
        raise NotImplementedError


class GreeterFactory(AbstractGreeterFactoryBase):
    def create(self, lang):
        if lang == "fr":
            return "bonjour"
        else:
            return "hello"


# Shell injection.
def run_user_command(user_input):
    return subprocess.call("echo " + user_input, shell=True)


# Mutable default argument.
def add_item(item, bucket=[]):
    bucket.append(item)
    return bucket


# Bare except swallows everything, including KeyboardInterrupt.
def load_config(path):
    try:
        f = open(path)
        data = json.loads(f.read())
        return data
    except:
        return None


# Non-idiomatic: index loop, manual counter, string concat in a loop.
def count_words(text):
    words = text.split(" ")
    counts = {}
    i = 0
    while i < len(words):
        w = words[i]
        if w in counts.keys():
            counts[w] = counts[w] + 1
        else:
            counts[w] = 1
        i = i + 1
    return counts


# This one is fine.
def add(a, b):
    return a + b


# chewgy-ignore-start
def intentionally_horrible():
    exec(os.environ.get("PAYLOAD", "pass"))
# chewgy-ignore-end


if __name__ == "__main__":
    print(GreeterFactory().create("fr"))
    print(add(1, 2))
    print(count_words("a b a"))

File: samples/messy.rs
// A deliberately awful Rust file for testing Chewgy.
// Open it in the Extension Development Host and hit Ctrl+Alt+C.

use std::collections::HashMap;

// Over-engineering: a trait and a factory for two hardcoded cases.
trait Greeter {
    fn greet(&self) -> String;
}

struct EnglishGreeter;
struct FrenchGreeter;

impl Greeter for EnglishGreeter {
    fn greet(&self) -> String {
        return String::from("hello");
    }
}

impl Greeter for FrenchGreeter {
    fn greet(&self) -> String {
        return String::from("bonjour");
    }
}

struct GreeterFactoryBuilder;

impl GreeterFactoryBuilder {
    fn build(lang: &str) -> Box<dyn Greeter> {
        if lang == "fr" {
            Box::new(FrenchGreeter)
        } else {
            Box::new(EnglishGreeter)
        }
    }
}

// Hardcoded secret.
const API_TOKEN: &str = "sk-live-do-not-actually-do-this-1234567890";

// Panics on any bad input.
fn parse_port(raw: &str) -> u16 {
    raw.parse::<u16>().unwrap()
}

// Off-by-one, plus an index that can panic.
fn last_item(items: &Vec<String>) -> String {
    let idx = items.len();
    items[idx].clone()
}

// Non-idiomatic: manual index loop, needless clone, needless collect.
fn count_words(text: String) -> HashMap<String, i32> {
    let words: Vec<String> = text.split(" ").map(|w| w.to_string()).collect();
    let mut counts: HashMap<String, i32> = HashMap::new();
    let mut i = 0;
    while i < words.len() {
        let word = words[i].clone();
        if counts.contains_key(&word) {
            let current = counts.get(&word).unwrap().clone();
            counts.insert(word, current + 1);
        } else {
            counts.insert(word, 1);
        }
        i = i + 1;
    }
    return counts;
}

// This one is fine. Chewgy should leave it alone.
fn add(a: i32, b: i32) -> i32 {
    a + b
}

// Chewgy is told to keep quiet about the next line.
// chewgy-ignore
fn deliberately_bad() -> i32 { std::fs::read_to_string("nope").unwrap().len() as i32 }

fn main() {
    println!("{}", GreeterFactoryBuilder::build("fr").greet());
    println!("{}", API_TOKEN);
    println!("{}", parse_port("8080"));
    println!("{}", add(1, 2));
    println!("{:?}", count_words("a b a".to_string()));
    println!("{:?}", last_item(&vec!["x".to_string()]));
    println!("{}", deliberately_bad());
}
